File: kernel/nodejs/src/bundle/files-integrity.ts
import { DEFAULT_MANIFEST_FILENAME, sha256Base64Url, type ArtifactLayer } from "@telorun/analyzer";
import * as fs from "fs/promises";
import * as path from "path";

import { findOwnerDoc, parseManifestDocs } from "./module-manifest.js";

/** A regular file of a payload layer; `executable` when it ships with its
 *  execute bit set. */
export interface PayloadRegularFile {
  /** POSIX-relative path inside the bundle. */
  name: string;
  content: Buffer | Uint8Array;
  executable?: boolean;
}

/** A symbolic link of a payload layer. `link` is the target exactly as the link
 *  stores it, relative to the link's own directory. */
export interface PayloadLink {
  /** POSIX-relative path inside the bundle. */
  name: string;
  link: string;
}

export type PayloadFile = PayloadRegularFile | PayloadLink;

/** A regular file known by its `sources:` pin rather than by bytes on disk — a
 *  staged file on a tree where nothing is staged. It digests exactly as the file
 *  it pins, and cannot be framed. */
export interface PayloadPinnedFile {
  /** POSIX-relative path inside the bundle. */
  name: string;
  /** Lowercase hex SHA-256 of the file's bytes, as the pin records it. */
  sha256: string;
  executable: boolean;
}

/** An entry a layer's integrity is computed over. */
export type LayerEntry = PayloadFile | PayloadPinnedFile;

export function isPayloadLink(file: LayerEntry): file is PayloadLink {
  return "link" in file;
}

export function isPinnedFile(file: LayerEntry): file is PayloadPinnedFile {
  return "sha256" in file;
}

/**
 * Canonical per-file content digest of one **layer**'s files — the `integrity`
 * value of that layer's entry in the published `layers:` index. SHA-256 over the
 * sorted lines of every file in the layer, `telo.yaml` excluded (it is the
 * manifest layer, which carries the index and so cannot hash itself; the
 * importer's `#sha256-...` pin covers it instead). The lines:
 *
 * - regular file: `<path>\0<sha256(content)>`
 * - executable file: `<path>\0<sha256(content)>\0x`
 * - symbolic link: `<path>\0l\0<target>`
 *
 * A regular file's line predates the other two and is unchanged, so every
 * integrity already published still verifies. The forms cannot collide: a path
 * holds no `\0`, and a digest is 43 base64url characters where `l` is one.
 *
 * Hashing file *contents* rather than the tar/gzip bytes makes the digest
 * independent of archive framing, so publisher and client compute the same
 * value from the same file set, and it can be re-derived from the extracted
 * files on disk ({@link readPayloadFile}) — which is what makes a per-layer
 * cache checkable without re-tarring, and why the execute bit and a link's
 * target are in the line. Distinct from the layer's `blob` digest, which covers
 * the pushed bytes and addresses the layer. Returns `sha256-<base64url>`.
 *
 * A pinned entry contributes the line of the file it pins, which is what lets a
 * tree with nothing staged compute the number publish derives from the bytes.
 */
export async function computeFilesIntegrity(files: readonly LayerEntry[]): Promise<string> {
  const lines: string[] = [];
  for (const file of files) {
    if (file.name === DEFAULT_MANIFEST_FILENAME) continue;
    if (isPayloadLink(file)) {
      lines.push(`${file.name}\0l\0${file.link}`);
      continue;
    }
    const digest = isPinnedFile(file)
      ? Buffer.from(file.sha256, "hex").toString("base64url")
      : await sha256Base64Url(
          file.content instanceof Uint8Array ? file.content : new Uint8Array(file.content),
        );
    lines.push(file.executable ? `${file.name}\0${digest}\0x` : `${file.name}\0${digest}`);
  }
  lines.sort();
  return `sha256-${await sha256Base64Url(new TextEncoder().encode(lines.join("\n")))}`;
}

/**
 * Read one payload file off disk as the kind it is: a symbolic link as a link
 * (never followed), a file with any execute bit as executable. What publish
 * reads a module directory with; re-deriving a materialized layer's integrity
 * from disk reads it the same way.
 */
export async function readPayloadFile(dir: string, name: string): Promise<PayloadFile> {
  const abs = path.resolve(dir, name);
  const stat = await fs.lstat(abs);
  if (stat.isSymbolicLink()) return { name, link: (await fs.readlink(abs)).replace(/\\/g, "/") };
  if (!stat.isFile()) {
    throw new Error(`'${name}' is neither a regular file nor a symbolic link, so it cannot ship.`);
  }
  const content = await fs.readFile(abs);
  return (stat.mode & 0o111) !== 0 ? { name, content, executable: true } : { name, content };
}

/**
 * Write the `layers:` index onto the manifest's owner doc so the published
 * `telo.yaml` pins and addresses every payload layer — transitively covered by
 * importers' `#sha256-...` hash over this manifest.
 *
 * Called after the payload blobs are pushed and before the manifest blob is,
 * which is what keeps the index non-circular: it names only layers other than
 * the one carrying it. Each layer's own digest excludes `telo.yaml`, so
 * injecting the index does not invalidate any of them. Returns the manifest
 * unchanged when it has no owner doc.
 */
export function injectLayerIndex(manifest: string, layers: readonly ArtifactLayer[]): string {
  const docs = parseManifestDocs(manifest);
  const owner = findOwnerDoc(docs);
  if (!owner) return manifest;
  owner.set(
    "layers",
    layers.map((layer) => ({
      role: layer.role,
      ...(layer.selector ? { selector: { ...layer.selector } } : {}),
      blob: layer.blob,
      integrity: layer.integrity,
    })),
  );
  return docs.map((d) => d.toString()).join("---\n");
}

File: kernel/rust/src/bundle/files_integrity.rs
//! A layer's file set and its content digest. Mirrors
//! `../../../nodejs/src/bundle/files-integrity.ts`.
//!
//! Read side only. `readPayloadFile` and `injectLayerIndex` serve publishing, and
//! the pinned entry a release digests a cold tree with never reaches a kernel, so
//! none of them has a counterpart here.

use telo_analyzer::sources::integrity::sha256_base64url;
use telo_analyzer::DEFAULT_MANIFEST_FILENAME;

/// One entry of a payload layer.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum PayloadFile {
    /// A regular file; `executable` when it ships with an execute bit set.
    Regular {
        /// POSIX-relative path inside the layer.
        name: String,
        content: Vec<u8>,
        executable: bool,
    },
    /// A symbolic link; `link` is the target exactly as the link stores it.
    Link { name: String, link: String },
}

impl PayloadFile {
    pub fn name(&self) -> &str {
        match self {
            PayloadFile::Regular { name, .. } | PayloadFile::Link { name, .. } => name,
        }
    }
}

/// The `integrity` of one layer (spec §3.3): SHA-256 over the sorted lines of
/// every entry, `telo.yaml` excluded, rendered `sha256-<base64url>`.
///
/// - regular file: `<path>\0<sha256(content)>`
/// - executable file: `<path>\0<sha256(content)>\0x`
/// - symbolic link: `<path>\0l\0<target>`
pub fn compute_files_integrity(files: &[PayloadFile]) -> String {
    let mut lines: Vec<String> = files
        .iter()
        .filter(|file| file.name() != DEFAULT_MANIFEST_FILENAME)
        .map(|file| match file {
            PayloadFile::Link { name, link } => format!("{name}\0l\0{link}"),
            PayloadFile::Regular {
                name,
                content,
                executable,
            } => {
                let digest = sha256_base64url(content);
                if *executable {
                    format!("{name}\0{digest}\0x")
                } else {
                    format!("{name}\0{digest}")
                }
            }
        })
        .collect();
    // By UTF-16 code unit, which is how the Node half's `Array.prototype.sort`
    // orders strings; byte order disagrees with it for a path outside the BMP.
    lines.sort_by(|a, b| a.encode_utf16().cmp(b.encode_utf16()));
    format!("sha256-{}", sha256_base64url(lines.join("\n").as_bytes()))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::Value;

    /// The `filesIntegrity` cases of the shared vector file, which
    /// `kernel/nodejs/tests/files-integrity-vectors.test.ts` runs too.
    #[test]
    fn agrees_with_the_shared_integrity_vectors() {
        let vectors: Value = serde_json::from_str(include_str!(
            "../../../../analyzer/artifact-axes/layer-index-vectors.json"
        ))
        .expect("vector file is JSON");
        for case in vectors["filesIntegrity"].as_array().expect("filesIntegrity cases") {
            let files: Vec<PayloadFile> = case["files"]
                .as_array()
                .unwrap()
                .iter()
                .map(|file| {
                    let name = file["name"].as_str().unwrap().to_string();
                    match file.get("link") {
                        Some(link) => PayloadFile::Link {
                            name,
                            link: link.as_str().unwrap().to_string(),
                        },
                        None => PayloadFile::Regular {
                            name,
                            content: file["content"].as_str().unwrap().as_bytes().to_vec(),
                            executable: file.get("executable") == Some(&Value::Bool(true)),
                        },
                    }
                })
                .collect();
            assert_eq!(
                compute_files_integrity(&files),
                case["integrity"].as_str().unwrap(),
                "{}",
                case["name"]
            );
        }
    }
}
